icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+   * @param line_sum - line sum to be copied because it is also given to triangle sum.
+   * @param bucket_ptr - Per to the bucket to be added to the line, passed as a Per to avoid copying.
+   * @param segment_idx - relevant segment to identify the returning result.
+   */
+  void set_phase2_line_addition(const P& line_sum, P* bucket_ptr, const int segment_idx)
+  {
+    m_p1 = line_sum;
+    m_p2_Per = bucket_ptr;
+    m_return_idx = segment_idx;
+    m_is_line = true;
+    m_p2_config = ADD_P1_AND_P2_PER;
+    dispatch();
+  }
+
+  /**
+   * @brief Set up triangle addition for phase 2.
+   * No need to specify segment idx as this is only assigned to a task already handling the triangle's segment idx
+   * (A previous line or triangle sum of this segment).
+   * @param line_sum - line sum to be copied because its value can be updated during the triangle sum execution.
+   * @param triangle_sum_ptr - Per to the triangle sum, passed as a Per to avoid unnecessary copying.
+   */
+  void set_phase2_triangle_addition(P& line_sum, P* triangle_sum_ptr)
+  {
+    m_p1 = line_sum;
+    m_p2_Per = triangle_sum_ptr;
+    m_is_line = false;
+    m_p2_config = ADD_P1_AND_P2_PER;
+    dispatch();
+  }
+
+  /**
+   * @brief Chain addition used in phase 1 when a collision between a new result and an occupied bucket.
+   * @param bucket - the bucket value to be added to the existing result in p1.
+   */
+  void set_phase1_collision_task(P& bucket)
+  {
+    m_p2 = bucket;
+    m_p2_config = ADD_P1_P2_BY_VALUE;
+    dispatch();
+  }
+
+  P m_p1;                 // One of the addends, and holds the addition result afterwards
+  int m_return_idx;       // Idx allowing manager to figure out where the result belong to.
+  bool m_is_line = false; // Indicator for phase 2 sums between line sum and triangle sum.
+
+private:
+  enum eAddType { ADD_P1_P2_BY_VALUE, ADD_P1_AND_P2_PER, ADD_P1_AND_AFFINE_P1 };
+  eAddType m_p2_config = ADD_P1_P2_BY_VALUE;
+
+  // Various configs of the second addend p2, one for each eAddType
+  P m_p2;
+  P const* m_p2_Per;
+  A m_p2_affine;
+};
+
+/**
+ * @class MSM
+ * @brief class for solving multi-scalar-multiplication on the cpu.
+ * The class is a template depending on the element relevant to MSM (for instance EC P).
+ * NOTE The msm runs only if nof_threads * 4 > nof_bms. The user can guarantee the condition by assigning a proper
+ * precompute factor that will decrease the number of BMs required to calculate the MSM.
+ */
+template <typename A, typename P>
+class Msm
+{
+public:
+  /**
+   * @brief Constructor for Msm class.
+   * @param config - msm config. important parameters that are part of the config extension are: n_threads, c
+   */
+  Msm(const MSMConfig& config);
+
+  /**
+   * @brief Destructor for Msm class.
+   * Ensures phase 3 threads have finished and joins them (other deletion are implemented implicitly).
+   */
+  ~Msm()
+  {
+    for (std::thread& p3_thread : m_p3_threads) {
```

Comment:
if p is phase then use the word phase. It makes code hard to read when have to guess what letters stand for. 
Even if longer, it is easier and shorter to read.

Reply:
done